File: ws/silica.py
import os
import uuid
import re
import subprocess
import argparse
import json
from subprocess import call
from flask import Flask, send_file, flash, render_template, request, redirect, url_for
from werkzeug.utils import secure_filename


SILICAWS = os.path.dirname(os.path.abspath(__file__))

app = Flask(__name__)
app.config['SILICA'] = os.path.join(SILICAWS, "..")
app.config['UPLOAD_FOLDER'] = os.path.join(app.config['SILICA'], "data")
app.config['MAX_CONTENT_LENGTH'] = 8 * 1024 * 1024   #maximum of 8MB
app.config['BASEURL'] = '/silica'


def allowed_file(filename):
   return '.' in filename and filename.rsplit('.', 1)[1].lower() in set(['fasta', 'fa', 'json', 'csv', 'txt'])

uuid_re = re.compile(r'(^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})-{0,1}([ap]{0,1})([cj]{0,1})$')
def is_valid_uuid(s):
   return uuid_re.match(s) is not None

@app.route('/download/<uuid>')
def download(uuid):
   if is_valid_uuid(uuid):
      ma = uuid_re.match(uuid)
      filename = "silica_" + ma.group(1)
      downName = "silica"
      if ma.group(2) == 'a':
         filename += "_amplicons"
         downName += "_amplicons"
      else:
         filename += "_primer"
         downName += "_primer"
      if ma.group(3) == 'c':
         filename += ".csv"
         downName += ".csv"
         miType = 'text/csv'
      else:
         filename += ".json"
         downName += ".json"
         miType = 'text/json'
      if allowed_file(filename):
         sf = os.path.join(app.config['UPLOAD_FOLDER'], uuid[0:2])
         if os.path.exists(sf):
            if os.path.isfile(os.path.join(sf, filename)):
               return send_file(os.path.join(sf, filename), mimetype=miType, as_attachment=True, attachment_filename=downName)
   return "File does not exist!"

@app.route('/results/<uuid>', methods = ['GET', 'POST'])
def results(uuid):
   amplicons = []
   primers = []
   showAmp = 0
   showPri = 0
   amphtml = ''
   primhtml = ''
   prStart = 0;
   amStart = 0;
   step = 30;
   moreAmp = 0;
   morePri = 0;

   if request.method == 'POST':
      if request.form['submit'] == 'Back to Submit Form':
         return redirect(app.config['BASEURL'] + "/upload", code=302)
      prStart = int(onlyInt(request.form['prStart']))
      amStart = int(onlyInt(request.form['amStart']))
      if request.form['submit'] == 'Amp Up':
         amStart -= step
      if request.form['submit'] == 'Amp Down':
         amStart += step
      if request.form['submit'] == 'Prim Up':
         prStart -= step
         showPri = 1
      if request.form['submit'] == 'Prim Down':
         prStart += step
         showPri = 1

   if is_valid_uuid(uuid):
      sf = os.path.join(app.config['UPLOAD_FOLDER'], uuid[0:2])
      if os.path.exists(sf):
         ampfilename = "silica_" + uuid + "_amplicons.json";
         ampdlink = '<a href="' + app.config['BASEURL'] + '/download/' + uuid + '-ac">CSV</a> or <a href="' + app.config['BASEURL'] + '/download/' + uuid + '-aj">JSON</a>'
         if allowed_file(ampfilename):
            if os.path.isfile(os.path.join(sf, ampfilename)):
               amplicons = json.loads(open(os.path.join(sf, ampfilename)).read())
         primfilename = "silica_" + uuid + "_primer.json";
         primdlink = '<a href="' + app.config['BASEURL'] + '/download/' + uuid + '-pc">CSV</a> or <a href="' + app.config['BASEURL'] + '/download/' + uuid + '-pj">JSON</a>'
         if allowed_file(primfilename):
            if os.path.isfile(os.path.join(sf, primfilename)):
               primers = json.loads(open(os.path.join(sf, primfilename)).read())
   if len(amplicons) > 0:
      if showPri == 0:
         showAmp = 1
      if amStart + step < len(amplicons):
         moreAmp = 1
         lastAmp = amStart + step
      else:
         lastAmp = len(amplicons)
      for a in range(amStart, lastAmp):
         amphtml += '<h3>Amplicon ' + str(amplicons[a]['Id'] + 1) +'</h3>\n<p>'
         amphtml += '<strong>Length:</strong> ' + str(amplicons[a]['Length']) +' bp<br />\n'
         amphtml += '<strong>Penalty:</strong> ' + str(amplicons[a]['Penalty']) +'<br />\n'
         amphtml += '<strong>Location:</strong> ' + str(amplicons[a]['Chrom']) + ':' + str(amplicons[a]['ForPos']) + '-' + str(amplicons[a]['RevPos'])+'<br />\n'
         amphtml += '<strong>Forward Primer Name:</strong> ' + str(amplicons[a]['ForName']) +'<br />\n'
         amphtml += '<strong>Forward Primer Tm:</strong> ' + str(amplicons[a]['ForTm']) +'&deg;C<br />\n'
         amphtml += '<strong>Forward Primer Sequence:</strong> ' + str(amplicons[a]['ForSeq']) +'<br />\n'
         amphtml += '<strong>Reverse Primer Name:</strong> ' + str(amplicons[a]['RevName']) +'<br />\n'
         amphtml += '<strong>Reverse Primer Tm:</strong> ' + str(amplicons[a]['RevTm']) +'&deg;C<br />\n'
         amphtml += '<strong>Reverse Primer Sequence:</strong> ' + str(amplicons[a]['RevSeq']) +'<br />\n'
         amphtml += '<strong>Amplicon Sequence:</strong><br /><pre>'
         splitSeq = str(amplicons[a]['Seq'])
         for i in range(0, len(splitSeq)):
            if i % 60 == 0 and i != 0:
               amphtml += '<br />'
            amphtml += splitSeq[i] 
         amphtml += '</pre><br />\n'
         amphtml += '</p>'
 
   if len(primers) > 0:
      if prStart + step < len(primers):
         morePri = 1
         lastPri = prStart + step
      else:
         lastPri = len(primers)
      for a in range(prStart, lastPri):
         primhtml += '<h3>Primer Binding Site ' + str(primers[a]['Id'] + 1) +'</h3>\n<p>'
         primhtml += '<strong>Primer Tm:</strong> ' + str(primers[a]['Tm']) +'&deg;C<br />\n'
         if str(primers[a]['Ori']) == 'reverse':
            startPos = int(primers[a]['Pos']) - len(str(primers[a]['Seq']))
            endPos = str(primers[a]['Pos'])
            loc = ' on reverse'
         else:
            startPos = str(primers[a]['Pos'])
            endPos = int(primers[a]['Pos']) + len(str(primers[a]['Seq']))
            loc = ' on forward'
         primhtml += '<strong>Location:</strong> ' + str(primers[a]['Chrom']) + ':' + str(startPos) + '-' + str(endPos) + loc + '<br />\n'
         primhtml += '<strong>Primer Name:</strong> ' + str(primers[a]['Name']) +'<br />\n'
         primhtml += '<strong>Primer Sequence:</strong> ' + str(primers[a]['Seq']) +'<br />\n'
         primhtml += '<strong>Genome Sequence:</strong> ' + str(primers[a]['Genome']) +'<br />\n'
         primhtml += '</p>'

   return render_template('results.html', baseurl = app.config['BASEURL']+ "/results/" + uuid, uuid=uuid, 
                          ampcount=len(amplicons), primecount=len(primers), showAmp=showAmp,
                          amphtml=amphtml, primhtml=primhtml, prStart=prStart, amStart=amStart,
                          moreAmp=moreAmp, morePri=morePri, ampdlink=ampdlink, primdlink=primdlink )
   return "File does not exist!"

@app.route('/upload', methods = ['GET', 'POST'])
def upload_file():
   if request.method == 'POST':
      uuidstr = str(uuid.uuid4())

      # Get subfolder
      sf = os.path.join(app.config['UPLOAD_FOLDER'], uuidstr[0:2])
      if not os.path.exists(sf):
         os.makedirs(sf)

      # Load Test Data
      if request.form['submit'] == 'Load Example Data':
         testData = '>FGA_f\nGCCCCATAGGTTTTGAACTCA\n>FGA_r\nTGATTTGTCTGTAATTGCCAGC'
         return render_template('upload.html', baseurl = app.config['BASEURL'], example = testData)

      # Fasta file
      primerData = request.form['fastaText']
      primerFile = 'direct.fa'
      if 'fasta' in request.files:
         ffa = request.files['fasta']
         if ffa.filename != '':
            if allowed_file(ffa.filename):
               primerData += '\n' + ffa.read()
               primerFile = secure_filename(ffa.filename)
      if  primerData == '':
         error = "Please provide a set of primers!"
         return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
      ffaname = os.path.join(sf, "silica_" + uuidstr + "_" + primerFile)
      primerData = primerData.replace('\r\n','\n')
      with open(ffaname, "w") as primFile:
         primFile.write(primerData)
    #  ffa.save(ffaname)

      # Genome
      val = request.form.get("submit", "None provided")
      genome = request.form['genome']
      if genome == '':
         error = "Please select a genome!"
         return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
      genome = os.path.join(app.config['SILICA'], "fm", genome)

      # Run Rscript
      outfile = os.path.join(sf, "silica_" + uuidstr + "_amplicons")
      prfile = os.path.join(sf, "silica_" + uuidstr + "_primer")
      paramfile = os.path.join(sf, "silica_" + uuidstr + "_parameter.txt")
      logfile = os.path.join(sf, "silica_" + uuidstr + ".log")
      errfile = os.path.join(sf, "silica_" + uuidstr + ".err")
      with open(logfile, "w") as log:
         with open(errfile, "w") as err:
            with open(paramfile, "w") as param:
               param.write("genome=" + genome + '\n')
               setAmpSize = onlyInt(request.form['setAmpSize'])
               param.write("maxProdSize=" + setAmpSize + '\n')
               setTmCutoff = onlyFloat(request.form['setTmCutoff'])
               param.write("cutTemp=" + setTmCutoff + '\n')
               if float(setTmCutoff) < 30.0:
                  error = 'Mimnimal Primer Tm must be >= 30&deg;C'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setKmer = onlyInt(request.form['setKmer'])
               param.write("kmer=" + setKmer + '\n')
               if int(setKmer) < 15:
                  error = 'Number of bp Used to Seach for Matches must be > 14 bp'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setEDis = onlyInt(request.form['setDist'])
               param.write("distance=" + setEDis + '\n')
               if int(setEDis) != 0 and int(setEDis) != 1:
                  error = 'Maximal Allowed Number of Mutations must be 0 or 1'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setCutoffPen = onlyFloat(request.form['setCutoffPen'])
               param.write("cutoffPenalty=" + setCutoffPen + '\n')
               if float(setCutoffPen) < 0.0 and int(setCutoffPen) != -1:
                  error = 'Keep Only PCR Products with Penalty Below must be > 0.0 or -1'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setPenTmDiff = onlyFloat(request.form['setPenTmDiff'])
               param.write("penaltyTmDiff=" + setPenTmDiff + '\n')
               if float(setPenTmDiff) < 0.0:
                  error = 'Penalty Factor for Single Primer Tm Mismatch must be >= 0.0'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setPenTmMismatch = onlyFloat(request.form['setPenTmMismatch'])
               param.write("penaltyTmMismatch=" + setPenTmMismatch + '\n')
               if float(setPenTmMismatch) < 0.0:
                  error = 'Penalty Factor for Tm Mismatch of Primers in a Pair must be >= 0.0'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setPenLength = onlyFloat(request.form['setPenLength'])
               param.write("penaltyLength=" + setPenLength + '\n')
               if float(setPenLength) < 0.0:
                  error = 'Penalty Factor for PCR Product Length must be >= 0.0'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setCtmMv = onlyFloat(request.form['setCtmMv'])
               param.write("monovalent=" + setCtmMv + '\n')
               if float(setCtmMv) < 0.0:
                  error = 'Concentration of Monovalent Ions must be >= 0.0 mMol'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setCtmDv = onlyFloat(request.form['setCtmDv'])
               param.write("divalent=" + setCtmDv + '\n')
               if float(setCtmDv) < 0.0:
                  error = 'Concentration of Divalent Ions must be >= 0.0 mMol'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setCtmDNA = onlyFloat(request.form['setCtmDNA'])
               param.write("dna=" + setCtmDNA + '\n')
               if float(setCtmDNA) < 0.1:
                  error = 'Concentration of Annealing(!) Oligos must be >= 0.1 nMol'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)
               setCtmDNTP = onlyFloat(request.form['setCtmDNTP'])
               param.write("dntp=" + setCtmDNTP + '\n')
               if float(setCtmDNTP) < 0.0:
                  error = 'Concentration of the Sum of All dNTPs must be >= 0.0 mMol'
                  return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)

               slexe = os.path.join(app.config['SILICA'], "./src/silica")
               return_code = call([slexe, '-g', genome, '-o', outfile, '-p', prfile,
                                          '--maxProdSize', setAmpSize, '--cutTemp', setTmCutoff,
                                          '--kmer', setKmer, '--distance', setEDis,
                                          '--cutoffPenalty', setCutoffPen, '--penaltyTmDiff', setPenTmDiff,
                                          '--penaltyTmMismatch', setPenTmMismatch, '--penaltyLength', setPenLength,
                                          '--monovalent', setCtmMv, '--divalent', setCtmDv,
                                          '--dna', setCtmDNA, '--dntp', setCtmDNTP,
                                          '-f', 'jsoncsv',
                                          ffaname], stdout=log, stderr=err)
      if return_code != 0:
         error = "Error in running Silica!"
         return render_template('upload.html', baseurl = app.config['BASEURL'], error = error)

      # Send download pdf
      return redirect(app.config['BASEURL'] + "/results/" + uuidstr, code=302)
   return render_template('upload.html', baseurl = app.config['BASEURL'])

@app.route("/")
def submit():
    return render_template("upload.html", baseurl = app.config['BASEURL'])

def onlyFloat(txt):
    onlyNumbDC = re.compile('[^0-9,.\-]')
    txt = onlyNumbDC.sub( '', txt)
    txt = txt.replace(',', '.')
    return txt

def onlyInt(txt):
    onlyNumb = re.compile('[^0-9\-]')
    return onlyNumb.sub( '', txt)

if __name__ == '__main__':
   parser = argparse.ArgumentParser(description='Silica App')
   parser.add_argument('-b', '--baseurl', type=str, required=False, default="", metavar="", dest='baseurl', help='baseurl')
   parser.add_argument('-d', '--debug', required=False, default=False, action='store_true', dest='debug', help='debug mode')
   parser.add_argument('-t', '--threaded', required=False, default=False, action='store_true', dest='threaded', help='threaded')
   parser.add_argument('-s', '--host', type=str, required=False, default="0.0.0.0", metavar="0.0.0.0", dest='host', help='host')
   parser.add_argument('-p', '--port', type=int, required=False, default=3300, metavar="3300", dest='port', help='port')
   args = parser.parse_args()
   app.config['BASEURL'] = args.baseurl
   app.run(host = args.host, port = args.port, debug = args.debug, threaded= args.threaded)

File: client/static/js/silica.js
/* global XMLHttpRequest */

var res = ""
var prStart = 0
var amStart = 0
var step = 30

var fileLoad = document.getElementById('fasta');
fileLoad.addEventListener('change', loadFasta, false);
var submitButton = document.getElementById('btn-submit')
submitButton.addEventListener('click', submit)
var sampleButton = document.getElementById('btn-example')
sampleButton.addEventListener('click', sampleData)
var helpButton = document.getElementById('btn-help')
helpButton.addEventListener('click', goToHelp)
const resultLink = document.getElementById('link-results')
const helpLink = document.getElementById('link-help')
const primerLink = document.getElementById('link-primers')
var sectionResults = document.getElementById('results')
var sectionAmplicons = document.getElementById('res-amplicons')
var sectionPrimers = document.getElementById('res-primers')

$('#mainTab a').on('click', function (e) {
  e.preventDefault()
  $(this).tab('show')
})

$('#resTab a').on('click', function (e) {
  e.preventDefault()
  $(this).tab('show')
})

var spinnerHtml = '<p>Analysis takes a couple of seconds to run, please be patient.</p><i class="fa fa-spinner fa-pulse fa-3x fa-fw"></i><br /><br />'

function checkPath(){
    var path = window.location.pathname
    if (path != "/") {
        loadLink (path)
    }
}

function goToHelp() {
    helpLink.click();
}

function loadFasta(f) {
    var file = f.target.files[0];
    if (file) { // && file.type.match("text/*")) {
        var reader = new FileReader();
        reader.onload = function(event) {
            var txt = event.target.result;
            var regEx1 = /\r\n/g;
            txt = txt.replace(regEx1, "\n");
            document.getElementById('fastaText').value = txt
        }
        reader.readAsText(file);
    } else {
        alert("Error opening file");
    }
}

function sampleData () {
    var fasta = document.getElementById('fastaText')
    fasta.value = '>FGA_f\nGCCCCATAGGTTTTGAACTCA\n>FGA_r\nTGATTTGTCTGTAATTGCCAGC\n';
    document.getElementById('genome').value = 'Homo_sapiens.GRCh37.dna.primary_assembly.fa.gz';
    document.getElementById('btn-genome').innerHTML = 'Homo Sapiens GRCh37';
}

function submit () {
    var data = new FormData();
    data.append('fastaText', document.getElementById('fastaText').value);
    data.append('genome', document.getElementById('genome').value);
    data.append('setAmpSize', document.getElementById('setAmpSize').value);
    data.append('setTmCutoff', document.getElementById('setTmCutoff').value);
    data.append('setKmer', document.getElementById('setKmer').value);
    data.append('setDist', document.getElementById('setDist').value);
    data.append('setCutoffPen', document.getElementById('setCutoffPen').value);
    data.append('setPenTmDiff', document.getElementById('setPenTmDiff').value);
    data.append('setPenTmMismatch', document.getElementById('setPenTmMismatch').value);
    data.append('setPenLength', document.getElementById('setPenLength').value);
    data.append('setCtmMv', document.getElementById('setCtmMv').value);
    data.append('setCtmDv', document.getElementById('setCtmDv').value);
    data.append('setCtmDNA', document.getElementById('setCtmDNA').value);
    data.append('setCtmDNTP', document.getElementById('setCtmDNTP').value);
    doSubmit (data);
}

function loadLink (uuid) {
    resultLink.click();
    var loca = 'http://0.0.0.0:3300';
    if (location.origin.startsWith("http")) {
        loca = location.origin;
    }
    var req = new XMLHttpRequest()
    req.addEventListener('load', displayResults)
    req.open('GET', loca + '/results' + uuid, true)
    req.send()
    sectionResults.innerHTML = spinnerHtml
}

function doSubmit (data) {
    resultLink.click();
    var loca = 'http://0.0.0.0:3300';
    if (location.origin.startsWith("http")) {
        loca = location.origin;
    }
    var req = new XMLHttpRequest()
    req.addEventListener('load', displayResults)
    req.open('POST', loca + '/upload', true)
    req.send(data)
    sectionResults.innerHTML = spinnerHtml
}

function displayResults() {
    if (this.status === 200) {
        displayData(this.response)
    } else {
        displayError(this.response)
    }
    resultLink.click();
}

function displayData(data) {
    res = JSON.parse(data)
    updateResults()
}

function displayError(data) {
    alert(data)
    var res = JSON.parse(data)
    for (var i = 0; i < res["errors"].length; i++) {
        sectionResults.innerHTML = '<br /><div class="error">' + res["errors"][i]['title'] + '</div><br />'
    }
    sectionAmplicons.innerHTML = '<br />'
    sectionPrimers.innerHTML = '<br />'
}

function updateResults() {
    var rHTML = ""
    var ampcount = res.data.amplicon.length
    var primecount = res.data.primer.length
    if (ampcount == 0) {
        rHTML += '<div class="alert alert-warning" role="alert"><strong>No Amplicons Found!</strong></div>\n'
	primerLink.click()
    }
    else {
        rHTML += '<div class="alert alert-success" role="alert"><strong> ' + ampcount + ' Amplicons Found!</strong></div>\n'
    }
    if (primecount == 0) {
        rHTML += '<div class="alert alert-warning" role="alert"><strong>No Primer Binding Sites Found!</strong></div>\n'
    }
    else {
        rHTML += '<div class="alert alert-success" role="alert"><strong> ' + primecount + ' Primer Binding Sites Found!</strong></div>\n'
    }
    sectionResults.innerHTML = '<br />' + rHTML + '<br />'
    rHTML = ""
    if (ampcount > 0) {
        rHTML += '<p>Download all as <a href="/silica/download' + res.data.link + '-ac">CSV</a>'
        rHTML += ' or <a href="/silica/download/' + res.data.link + '-aj">JSON</a></p>'
        if (amStart != 0) {
            rHTML += '<button type="button" class="btn btn-secondary" onclick="ampUp()">Up</button>'
        }
        var amStop = amStart + step
        var moreAmp = 1
        if (amStop > ampcount) {
            moreAmp = 0
            amStop = ampcount
        }
        for (var i = amStart; i < amStop; i++) {
            var amp = res.data.amplicon[i]
            rHTML += '<h3>Amplicon ' + (parseInt(amp['Id']) + 1) +'</h3>\n<p>'
            rHTML += '<strong>Length:</strong> ' + amp['Length'] +' bp<br />\n'
            rHTML += '<strong>Penalty:</strong> ' + amp['Penalty'] +'<br />\n'
            rHTML += '<strong>Location:</strong> ' + amp['Chrom'] + ':' + amp['ForPos'] + '-' + amp['RevPos']+'<br />\n'
            rHTML += '<strong>Forward Primer Name:</strong> ' + amp['ForName'] +'<br />\n'
            rHTML += '<strong>Forward Primer Tm:</strong> ' + amp['ForTm'] +'&deg;C<br />\n'
            rHTML += '<strong>Forward Primer Sequence:</strong> ' + amp['ForSeq'] +'<br />\n'
            rHTML += '<strong>Reverse Primer Name:</strong> ' + amp['RevName'] +'<br />\n'
            rHTML += '<strong>Reverse Primer Tm:</strong> ' + amp['RevTm'] +'&deg;C<br />\n'
            rHTML += '<strong>Reverse Primer Sequence:</strong> ' + amp['RevSeq'] +'<br />\n'
            rHTML += '<strong>Amplicon Sequence:</strong><br /><pre>'
            var splitSeq = amp['Seq']
            for (var k = 0; k < splitSeq.length ; k++) {
                if((k % 60 == 0) && (k != 0)) {
                    rHTML += '<br />'
                }
                rHTML += splitSeq.charAt(k)
            }
            rHTML += '</pre><br />\n'
            rHTML += '</p>'
	}
        if (moreAmp == 1) {
            rHTML += '<button type="button" class="btn btn-secondary" onclick="ampDown()">Down</button>'
        }
    }
    sectionAmplicons.innerHTML = '<br />' + rHTML + '<br />'
    rHTML = ""
    if (primecount > 0) {
        rHTML += '<p>Download all as <a href="/silica/download' + res.data.link + '-pc">CSV</a>'
        rHTML += ' or <a href="/silica/download/' + res.data.link + '-pj">JSON</a></p>'
        if (prStart != 0) {
            rHTML += '<button type="button" class="btn btn-secondary" onclick="primerUp()">Up</button>'
        }
        var prStop = prStart + step
        var morePrim = 1
        if (prStop > primecount) {
            morePrim = 0
            prStop = primecount
        }
        for (var i = prStart; i < prStop; i++) {
            var prim = res.data.primer[i]
            rHTML += '<h3>Primer Binding Site ' + (parseInt(prim['Id']) + 1) +'</h3>\n<p>'
            rHTML += '<strong>Primer Tm:</strong> ' + prim['Tm'] +'&deg;C<br />\n'
            if (prim['Ori'] == 'reverse') {
                rHTML += '<strong>Location:</strong> ' + prim['Chrom'] + ':' + (parseInt(prim['Pos']) - prim['Seq'].length) + '-' + prim['Pos'] + ' on reverse<br />\n'
            } else {
                rHTML += '<strong>Location:</strong> ' + prim['Chrom'] + ':' + prim['Pos'] + '-' + (parseInt(prim['Pos']) + prim['Seq'].length) + ' on forward<br />\n'
            }
            rHTML += '<strong>Primer Name:</strong> ' + prim['Name'] +'<br />\n'
            rHTML += '<strong>Primer Sequence:</strong> ' + prim['Seq'] +'<br />\n'
            rHTML += '<strong>Genome Sequence:</strong> ' + prim['Genome'] +'<br />\n'
            rHTML += '</p>'
        }
        if (morePrim == 1) {
            rHTML += '<button type="button" class="btn btn-secondary" onclick="primerDown()">Down</button>'
	}
    }
    sectionPrimers.innerHTML = '<br />' + rHTML + '<br />'
}

function ampUp() {
    amStart = amStart - step
    updateResults()
}

function ampDown() {
    amStart = amStart + step
    updateResults()
}

function primerUp() {
    prStart = prStart - step
    updateResults()
}

function primerDown() {
    prStart = prStart + step
    updateResults()
}

